contracts/sol6/fprConversionRates/QtyStepConversionRates.sol

```diff
@@ -0,0 +1,436 @@
+pragma solidity 0.6.6;
+
+import "../IERC20.sol";
+import "../utils/Utils5.sol";
+import "../IConversionRates.sol";
+import "./SimpleVolumeImbalanceRecorder.sol";
+
+contract QtyStepConversionRates is IConversionRates, SimpleVolumeImbalanceRecorder, Utils5 {
+    int256 internal constant MAX_IMBALANCE = 2**255 - 1;
+    uint256 internal constant POW_2_128 = 2**128;
+    int256 internal constant MAX_BPS_ADJUSTMENT = 10**11; // 1B %
+    int256 internal constant MIN_BPS_ADJUSTMENT = -100 * 100; // cannot go down by more than 100%
+    // step data constant
+    uint256 internal constant MAX_STEPS_IN_FUNCTION = 10;
+    int128 internal constant MAX_STEP_VALUE = 2**127 - 1;
+    int128 internal constant MIN_STEP_VALUE = -1 * 2**127;
+    // compact data constant
+    uint256 internal constant NUM_TOKENS_IN_COMPACT_DATA = 14;
+    uint256 internal constant BYTES_14_OFFSET = (2**(8 * NUM_TOKENS_IN_COMPACT_DATA));
+
+    // bps - basic rate steps. one step is 1 / 10000 of the rate.
+    // struct StepFunction {
+    //     int256[] x; // quantity for each step. Quantity of each step includes previous steps.
+    //     int256[] y; // rate change per quantity step  in bps.
+    // }
+    // 1 word
+    struct TokenData {
+        bool listed; // was added to reserve
+        bool enabled; // whether trade is enabled
+        // position in the compact data
+        // should we compress two things into 1 word
+        uint16 compactDataArrayIndex;
+        uint16 compactDataFieldIndex;
+        // rate data. base and changes according to quantity and reserve balance.
+        // generally speaking. Sell rate is 1 / buy rate i.e. the buy in the other direction.
+
+        // rate <= max_rate = 10^25 < 2^104 (10^31)
+        uint104 baseBuyRate; // in PRECISION units. see KyberConstants
+        uint104 baseSellRate; // PRECISION units. without (sell / buy) spread it is 1 / baseBuyRate
+    }
+
+    struct StepFuncions {
+        int256[] buyRateQtyStepFunction;
+        int256[] sellRateQtyStepFunction;
+    }
+
+    /*
+    this is the data for tokenRatesCompactData
+    but solidity compiler optimizer is sub-optimal, 
```

Comment:
this is not true - actually it wasn't true for 0.4.18 as well. but we discovered that in a later stage.

the compiler should handle this correctly.
so we should remove this comment and use this structure directly in storage
and of course check gas has not changed

Reply:
If we use this as array of struct, solidity compiler would not optimize this.

So do you want to change tokenRatesCompactData to mapping?

Reply:
https://gist.github.com/ducquangkstn/8a716879b012e4d5d71c9e9095b2b9e3

Reply:
update at e942b0a